nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractConfiguredComponent.java

```diff
@@ -90,47 +107,80 @@ public void setAnnotationData(final String data) {
-                                oldNode.removeReference(this);
-                            }
-                        }
-
-                        final ControllerServiceNode newNode = serviceProvider.getControllerServiceNode(value);
-                        if (newNode != null) {
-                            newNode.addReference(this);
+            try (final NarCloseable narCloseable = NarCloseable.withComponentNarLoader(component.getClass(), id)) {
+                final Set<String> modulePaths = new LinkedHashSet<>();
+                for (final Map.Entry<String, String> entry : properties.entrySet()) {
+                    if (entry.getKey() != null && entry.getValue() == null) {
+                        removeProperty(entry.getKey());
+                    } else if (entry.getKey() != null) {
+                        setProperty(entry.getKey(), entry.getValue());
+
+                        // for any properties that dynamically modify the classpath, attempt to evaluate them for expression language
+                        final PropertyDescriptor descriptor = component.getPropertyDescriptor(entry.getKey());
+                        if (descriptor.isDynamicClasspathModifier() && !StringUtils.isEmpty(entry.getValue())) {
+                            final StandardPropertyValue propertyValue = new StandardPropertyValue(entry.getValue(), null, variableRegistry);
+                            modulePaths.add(propertyValue.evaluateAttributeExpressions().getValue());
                         }
                     }
+                }
 
-                    try {
-                        component.onPropertyModified(descriptor, oldValue, value);
-                    } catch (final Exception e) {
-                        // nothing really to do here...
-                    }
+                final boolean requiresInstanceClassLoading = ExtensionManager.requiresInstanceClassLoading(component.getClass().getTypeName());
+                if (requiresInstanceClassLoading) {
+                    processClasspathModifiers(modulePaths);
+                } else if (!requiresInstanceClassLoading && modulePaths.size() > 0) {
+                    // Component has property descriptors with dynamicallyModifiesClasspath set to true, but the class does not have @RequiresInstanceClassLoading
+                    logger.warn("One or more property descriptors intend to modify the classpath, " +
+                            "but component does not contain the {} annotation, classpath will not be modified",
+                            new Object[] {RequiresInstanceClassLoading.class.getName()});
                 }
             }
         } finally {
             lock.unlock();
         }
     }
 
+    // Keep setProperty/removeProperty private so that all calls go through setProperties
+    private void setProperty(final String name, final String value) {
+        if (null == name || null == value) {
+            throw new IllegalArgumentException();
+        }
```

Comment:
Hmm. . . No message in the exception? Also, should we even check for nulls here given that it's private? meaning we have full control of when/how/where it is invoked?